// mapi/lib/search/searchlinks.js
var names = require('../base/names.js'),
    links = require('../base/links.js'),
    mlm = new links.manager();

var dewikiArtistLinkOptions = {
    linkdata: {
        protocol: 'http',
        port: '',
        host: 'de.wikipedia.org',
        pathname: 'w/api.php',
        hash: '',
        query: {
            action: 'query',
            list: 'search',
            srsearch: '',
            format: 'json'
        }
    },
    name: names.DEWIKINAME,
    description: 'German Wikipedia Search',
    searchParameter: 'srsearch'
};

var enwikiArtistLinkOptions = {
    linkdata: {
        protocol: 'http',
        port: '',
        host: 'en.wikipedia.org',
        pathname: 'w/api.php',
        hash: '',
        query: {
            action: 'query',
            list: 'search',
            srsearch: '',
            format: 'json'
        }
    },
    name: names.ENWIKINAME,
    description: 'English Wikipedia Search',
    searchParameter: 'srsearch'
};

var musicbrainzArtistLinkOptions = {
    linkdata: {
        protocol: 'http',
        port: '',
        host: 'www.musicbrainz.org',
        pathname: 'ws/2/artist',
        hash: '',
        query: {
            query: '',
            fmt: 'json'
        }
    },
    name: names.MBNAME,
    description: 'Musicbrainz Artist Search',
    searchParameter: 'query'
};

var discogsArtistLink1Options = {
    linkdata: {
        protocol: 'http',
        port: '',
        host: 'api.discogs.com',
        pathname: 'artist/{{search}}',
        hash: '',
        query: {
            f: 'json'
        }
    },
    name: names.DISCOGSARTISTNAME,
    description: 'discogs Artist Search',
    searchParameter: '',
    searchIsParameter: false
};

var discogsArtistLink2Options = {
    linkdata: {
        protocol: 'http',
        port: '',
        host: 'api.discogs.com',
        pathname: 'database/search',
        hash: '',
        query: {
            q: '',
            per_page: '10',
            type: 'artist',
            f: 'json'
        }
    },
    name: names.DISCOGSDBNAME,
    description: 'discogs Database Search',
    searchParameter: 'q',
    searchIsParameter: true
};

mlm.setUrlBuilder(require('../base/urlbuilder.js').urlbuilder);

mlm.addLink(mlm.createLink(discogsArtistLink1Options));    
mlm.addLink(mlm.createLink(discogsArtistLink2Options));
mlm.addLink(mlm.createLink(musicbrainzArtistLinkOptions));
mlm.addLink(mlm.createLink(dewikiArtistLinkOptions));
mlm.addLink(mlm.createLink(enwikiArtistLinkOptions));

module.exports.maslm = mlm;
